### PR TITLE
chore!: upgrade fuel core to `0.28.0`

### LINKED ISSUE #2447: Unskip e2e tests
We need to update the hardcoded IDs and unskip the tests skipped in https://github.com/FuelLabs/fuels-ts/pull/2430 — https://github.com/FuelLabs/fuels-ts/pull/2430/commits/b98943f10a630035f52d8d04925808bcca1097ec.

### LINKED ISSUE #2473: Upgrade `fuel-core` to `0.28.0`
Changelog notes:
 - https://github.com/FuelLabs/fuel-core/releases/tag/v0.28.0

### PR DESCRIPTION
**Pull request by petertonysmith94**
- Close #2473
- Close #2447

### Breaking changes
- `FUEL_TESTNET_NETWORK_URL` changed to `TESTNET_NETWORK_URL`

### CODE REVIEW

File: apps/docs-snippets/src/guide/introduction/getting-started.test.ts [hunk 1]
```diff
@@ -11,9 +11,10 @@ import {
  * @group browser
  */
 describe('Getting started', () => {
-  beforeAll(() => {
+  beforeAll(async () => {
     // Avoids using the actual network.
-    vi.spyOn(Provider, 'create').mockImplementationOnce(() => Provider.create(FUEL_NETWORK_URL));
+    const mockProvider = await Provider.create(FUEL_NETWORK_URL);
+    vi.spyOn(Provider, 'create').mockResolvedValue(mockProvider);
```
Comment (arboleya) [inline apps/docs-snippets/src/guide/introduction/getting-started.test.ts:17]: Why did you change this?

Shouldn't the `mockResolvedValue()` method return a promise, as the mocked method `create` does?

File: packages/account/src/configs.test.ts [hunk 2]
```diff
@@ -7,6 +7,11 @@ describe('Configs', () => {
     expect(configs.FUEL_NETWORK_URL).toBe('http://127.0.0.1:4000/v1/graphql');
   });
 
+  it('exports FUEL_DEVNET_NETWORK_URL', async () => {
+    const configs = await import('./configs');
+    expect(configs.FUEL_DEVNET_NETWORK_URL).toBe('https://devnet.fuel.network/v1/graphql');
```
Comment (arboleya) [inline packages/account/src/configs.test.ts:12]: That's great!

Can we trim these constant names a bit?

 - `DEVNET_NETWORK_URL`
 - `TESTNET_NETWORK_URL`

File: packages/account/src/configs.ts [hunk 3]
```diff
@@ -3,4 +3,5 @@ export const FUEL_NETWORK_URL: string =
     ? process?.env?.FUEL_NETWORK_URL || 'http://127.0.0.1:4000/v1/graphql'
     : 'http://127.0.0.1:4000/v1/graphql';
 
+export const FUEL_DEVNET_NETWORK_URL: string = 'https://devnet.fuel.network/v1/graphql';
```
Comment (arboleya) [inline packages/account/src/configs.ts:6]: Ni - same for `FUEL_TESTNET_NETWORK_URL`.

```suggestion
export const FUEL_DEVNET_NETWORK_URL = 'https://devnet.fuel.network/v1/graphql';
```

File: packages/fuel-gauge/src/e2e-script.test.ts [hunk 4]
```diff
@@ -12,19 +29,22 @@ describe('Live Script Test', () => {
   const UPGRADE_TX_ID = '0xe2c03044fe708e9b112027881baf9f892e6b64a630a629998922c1cab918c094';
   const UPLOAD_TX_ID = '0x94bc2a189b8211796c8fe5b9c6b67624fe97d2007e104bf1b30739944f43bd73';
 
+  const selectedNetwork: keyof typeof networks = 'devnet';
```
Comment (arboleya) [inline packages/fuel-gauge/src/e2e-script.test.ts:32]: Should we use an array for this? Maybe an enum as well?

```ts
export enum NETWORKS {
  DEVNET = 'devnet',
  TESTNET = 'devnet',
}

const selectedNetworks: NETWORKS[] = [NETWORKS.DEVNET, NETWORKS.TESTNET];
```

With this, we can run e2e's selectively as we go, or even target both networks simultaneously.

Makes sense?

File: packages/fuel-gauge/src/e2e-script.test.ts [hunk 5]
```diff
@@ -47,8 +67,7 @@ describe('Live Script Test', () => {
       console.error((e as Error).message);
       console.warn(`
         not enough coins to fit the target?
-        - add assets: https://faucet-testnet.fuel.network/
-        - check balance: https://app.fuel.network/account/${address}/assets/
+        - add assets: https://faucet-${selectedNetwork}.fuel.network/
```
Comment (arboleya) [inline packages/fuel-gauge/src/e2e-script.test.ts:70]: Did you consider having a `faucetUrl` property for [network members](https://github.com/FuelLabs/fuels-ts/blob/89667c895825a4d3ee9c98f928d8b517dcfc66c1/packages/fuel-gauge/src/e2e-script.test.ts#L12-L21)?

**Review by arboleya — COMMENTED** [review]
Good stuff! 🚀

File: packages/fuel-gauge/src/e2e-script.test.ts (hunk 4 above)
Comment (arboleya) [reply packages/fuel-gauge/src/e2e-script.test.ts]: We should also define these TX IDs inside the Network object so they can be individualized per chain:

 - `MINT_TX_ID`
 - `UPGRADE_TX_ID`
 - `UPLOAD_TX_ID`

File: apps/docs-snippets/src/guide/introduction/getting-started.test.ts (hunk 1 above)
Comment (arboleya) [reply apps/docs-snippets/src/guide/introduction/getting-started.test.ts:11]: The rationale is that mocks should return the same value as the methods they're mocking.

I suppose the original `Provide.create` returns an _unresolved_ promise, and so should the mock. 🤔 


```ts
const provider = Provider.create(FUEL_NETWORK_URL); // <— no await here
vi.spyOn(Provider, 'create').mockResolvedValue(provider);

// or simply
vi.spyOn(Provider, 'create').mockResolvedValue(Provider.create(FUEL_NETWORK_URL));
```

File: packages/fuel-gauge/src/e2e-script.test.ts (hunk 4 above)
Comment (arboleya) [reply packages/fuel-gauge/src/e2e-script.test.ts]: @Torres-ssf Should we use the opportunity to grab some new IDs from `testnet`?

This could solve #2447 before achieving a final solution via #2471 and/or #2474.

File: apps/docs-snippets/src/guide/introduction/getting-started.test.ts [hunk 6]
```diff
@@ -1,19 +1,14 @@
-import {
-  FUEL_NETWORK_URL,
-  FUEL_TESTNET_NETWORK_URL,
-  Provider,
-  Wallet,
-  WalletUnlocked,
-} from 'fuels';
+import { FUEL_NETWORK_URL, TESTNET_NETWORK_URL, Provider, Wallet, WalletUnlocked } from 'fuels';
```
Comment (arboleya) [inline apps/docs-snippets/src/guide/introduction/getting-started.test.ts:1]: Do we still need `FUEL_NETWORKL_URL`? 🤔 

cc @Torres-ssf

File: packages/fuel-gauge/src/e2e-script.test.ts [hunk 7]
```diff
@@ -1,30 +1,72 @@
...
+    [TransactionType.Upload]: string;
+  };
+};
+
+const configuredNetworks = {
+  [Networks.DEVNET]: {
+    networkUrl: DEVNET_NETWORK_URL,
+    privateKey: process.env.DEVNET_WALLET_PVT_KEY,
+    faucetUrl: `https://faucet-devnet.fuel.network/`,
+    transactions: {
```
Comment (arboleya) [inline packages/fuel-gauge/src/e2e-script.test.ts:33]: Nit.

```suggestion
    txIds: {
```

File: apps/docs-snippets/src/guide/introduction/getting-started.test.ts (hunk 6 above)
Comment (Torres-ssf) [reply apps/docs-snippets/src/guide/introduction/getting-started.test.ts:1]: @arboleya the `FUEL_NETWORK_URL` constant is mainly ( if not only ) used within our tests, and its value is: `http://127.0.0.1:4000/v1/graphql`. 

@FuelLabs/sdk-ts Does anyone remember why we export this constant from our umbrella package?

File: packages/fuel-gauge/src/e2e-script.test.ts (hunk 4 above)
Comment (Torres-ssf) [reply packages/fuel-gauge/src/e2e-script.test.ts]: @arboleya @petertonysmith94 Getting IDs for `upgrade` and `upload` TXs will be only possible if any of these transactions were already submitted on the `testnet`.

File: packages/account/src/providers/fuel-core-schema.graphql [hunk 8]
```diff
@@ -559,6 +570,10 @@ type Header {
   applicationHash: Bytes32!
 }
 
+enum HeaderVersion {
+  V1
+}
+
```
Comment (Torres-ssf) [inline packages/account/src/providers/fuel-core-schema.graphql:576]: @petertonysmith94 Just to confirm, you have updated this file by running the script `pnpm build:schema` from the `account/package.json` file, right?

File: packages/account/test/fixtures/chain.ts [hunk 9]
```diff
@@ -139,6 +139,11 @@ export const MOCK_CHAIN: GqlChainInfoFragment = {
       wqmm: '3',
       xor: '1',
       xori: '1',
+      alocDependentCost: {
+        type: 'LightOperation',
+        base: '2',
+        unitsPerGas: '214',
+      },
```
Comment (Torres-ssf) [inline packages/account/test/fixtures/chain.ts:146]: Were these values obtained from the `devnet`?

File: apps/docs-snippets/src/guide/introduction/getting-started.test.ts (hunk 6 above)
Comment (arboleya) [reply apps/docs-snippets/src/guide/introduction/getting-started.test.ts:1]: @Torres-ssf I was thinking if we could go with something like this:
 - `LOCAL_NETWORK_URL`
 - `DEVNET_NETWORK_URL`
 - `TESTNET_NETWORK_URL`
 - `NETWORK_URL` (for upcoming mainnet)

Thoughts?

Comment (Torres-ssf) [reply apps/docs-snippets/src/guide/introduction/getting-started.test.ts:1]: @arboleya This is a good suggestion. So the `LOCAL_NETWORK_URL` will be `http://127.0.0.1:4000/v1/graphql` ?

File: .github/workflows/test.yaml [hunk 10]
```diff
@@ -79,10 +79,9 @@ jobs:
       - name: Run Isolated Tests
         run: pnpm test:e2e
         env:
-          FUEL_NETWORK_URL: https://testnet.fuel.network/v1/graphql
-          FUEL_TESTNET_NETWORK_URL: https://testnet.fuel.network/v1/graphql
-          TEST_WALLET_PVT_KEY: ${{ secrets.TEST_WALLET_PVT_KEY }}
-          TEST_WALLET_ADDRESS: ${{ secrets.TEST_WALLET_ADDRESS }}
+          SHOULD_SKIP_E2E: false
```
Comment (danielbate) [inline .github/workflows/test.yaml:82]:
```suggestion
```

Is this being used anywhere?

File: apps/docs-snippets/src/guide/introduction/getting-started.test.ts (hunk 6 above)
Comment (danielbate) [reply apps/docs-snippets/src/guide/introduction/getting-started.test.ts:1]: The one caveat is that when copying from snippets, `FUEL_NETWORK_URL` used a ternary that could evaluate but live and local networks, so it could be extracted from snippets. Whereas someone now will need to change the constant dependent on if they are working on testnet or locally.

File: packages/fuel-gauge/src/e2e-script.test.ts [hunk 11]
```diff
@@ -47,25 +98,57 @@ describe('Live Script Test', () => {
       console.error((e as Error).message);
       console.warn(`
         not enough coins to fit the target?
-        - add assets: https://faucet-testnet.fuel.network/
-        - check balance: https://app.fuel.network/account/${address}/assets/
```
Comment (danielbate) [inline packages/fuel-gauge/src/e2e-script.test.ts:51]: How come this was removed?

**Review by danielbate — DISMISSED** [review]

File: apps/docs-snippets/src/guide/introduction/getting-started.test.ts (hunk 6 above)
Comment (arboleya) [reply apps/docs-snippets/src/guide/introduction/getting-started.test.ts:1]: Good point. Do you think this is a problem tho? Or a feature?

I wonder if such a constant should be static and avoid the ternary.

File: .fuel-core/configs/chainConfig.json [hunk 12]
```diff
@@ -129,6 +129,12 @@
           "wqmm": 10,
           "xor": 2,
           "xori": 2,
+          "alloc": {
+            "LightOperation": {
+              "base": "2",
+              "unitsPerGas": "214"
+            }
+          },
```
Comment (Torres-ssf) [inline .fuel-core/configs/chainConfig.json:137]: Why are we using `alloc` here? It seems that the added gas cost instruction was `alocDependentCost`.

Also, it appears that this gas instruction is not set on the dev/test nets
Screenshot 2024-06-12 at 07 11 49

**Review by Torres-ssf — CHANGES_REQUESTED** [review]

File: .fuel-core/configs/chainConfig.json (hunk 12 above)
Comment (Torres-ssf) [reply .fuel-core/configs/chainConfig.json:137]: UPDATE:

`alocDependentCost` it is set on the `devnet`, but not on the `testnet`, very likely because it was not updated yet.

**Review by Torres-ssf — DISMISSED** [review]

File: apps/docs-snippets/src/guide/introduction/getting-started.test.ts (hunk 6 above)
Comment (arboleya) [reply apps/docs-snippets/src/guide/introduction/getting-started.test.ts:1]: > @arboleya should we have an issue for collating all the little chores for mainnet?
(such as adding network URL, chain ID, etc)

@petertonysmith94 Like an epic? Definitely.

**Review by danielbate — DISMISSED** [review]

**Review by nedsalk — DISMISSED** [review]

File: packages/fuel-gauge/src/e2e-script.test.ts [hunk 13]
```diff
@@ -1,29 +1,79 @@
...
+    txIds: {
+      [TransactionType.Upgrade]:
+        '0xd64e3f7589bc1c6dcf1e419f4a3a8fc21d3694abf98f151000f34682d1cacdce',
+      [TransactionType.Upload]:
+        '0x996eec87a702ac978663fe67dbde7ab94d31f32b1860fbfc527d4b5447b3446c',
+    },
+  } as ConfiguredNetwork,
+};
+
+const selectedNetworks: Networks[] = [Networks.DEVNET];
```
Comment (arboleya) [inline packages/fuel-gauge/src/e2e-script.test.ts:56]: According to what we discussed, the PR must pass against both networks before we merge it. Correct?

```suggestion
const selectedNetworks: Networks[] = [Networks.DEVNET, Networks.TESTNET];
```

**Review by arboleya — CHANGES_REQUESTED** [review]

**Review by arboleya — APPROVED** [review]

**Review by danielbate — APPROVED** [review]

**Review by nedsalk — APPROVED** [review]
